benchmark/bls.cpp

```diff
@@ -0,0 +1,417 @@
+      chain->execute_setup_policy( setup_policy::full );
+
+      // get hold controller
+      control = chain->control.get();
+
+      // create account and deploy contract for a temp transaction
+      chain->create_accounts( {"payloadless"_n} );
+      chain->set_code( "payloadless"_n, test_contracts::payloadless_wasm() );
+      chain->set_abi( "payloadless"_n, test_contracts::payloadless_abi() );
+
+      // construct a signed transaction
+      fc::variant pretty_trx = fc::mutable_variant_object()
+         ("actions", fc::variants({
+            fc::mutable_variant_object()
+               ("account", name("payloadless"_n))
+               ("name", "doit")
+               ("authorization", fc::variants({
+                  fc::mutable_variant_object()
+                     ("actor", name("payloadless"_n))
+                     ("permission", name(config::active_name))
+               }))
+               ("data", fc::mutable_variant_object()
+               )
+            })
+        );
+      trx = std::make_unique<signed_transaction>();
+      abi_serializer::from_variant(pretty_trx, *trx, chain->get_resolver(), abi_serializer::create_yield_function( chain->abi_serializer_max_time ));
+      chain->set_transaction_headers(*trx);
+      trx->sign( chain->get_private_key( "payloadless"_n, "active" ), control->get_chain_id() );
+
+      // construct a packed transaction
+      ptrx = std::make_unique<packed_transaction>(*trx, eosio::chain::packed_transaction::compression_type::zlib);
+
+      // build transaction context from the packed transaction
+      timer = std::make_unique<platform_timer>();
+      trx_timer = std::make_unique<transaction_checktime_timer>(*timer);
+      trx_ctx = std::make_unique<transaction_context>(*control, *ptrx, ptrx->id(), std::move(*trx_timer));
+      trx_ctx->max_transaction_time_subjective = fc::microseconds::maximum();
+      trx_ctx->init_for_input_trx( ptrx->get_unprunable_size(), ptrx->get_prunable_size() );
+      trx_ctx->exec();
+
+      // build apply context from the control and transaction context
+      apply_ctx = std::make_unique<apply_context>(*control, *trx_ctx, 1);
+
+      // finally build the interface
+      interface = std::make_unique<webassembly::interface>(*apply_ctx);
+   }
+
+   std::unique_ptr<tester>                      chain;
+   controller*                                  control;
+   std::unique_ptr<signed_transaction>          trx;
+   std::unique_ptr<packed_transaction>          ptrx;
+   std::unique_ptr<platform_timer>              timer;
+   std::unique_ptr<transaction_checktime_timer> trx_timer;
+   std::unique_ptr<transaction_context>         trx_ctx;
+   std::unique_ptr<apply_context>               apply_ctx;
+   std::unique_ptr<webassembly::interface>      interface;
+};
+} // namespace eosio::chain
+
+
+namespace benchmark {
+
+// utilility to create a random scalar
+std::array<uint64_t, 4> random_scalar()
```

Comment:
I wonder if this should use `bls12_381::secret_key`. Are the new enforcement of valid data going to be tripped by this?

Reply:
I think the way `g1` and `g2` are generated this benchmarking (multiple by `g1::one()` or `g2::one()`) makes the resulting `g1` and `g2` are on the curve. Current `bls_pairing`, `bls_g1_map`, and `bls_g2_map` already verify input, and I tried g1/g2 add with validation enabled. They all work.